virttest/qemu_devices/qcontainer.py

```diff
@@ -1202,6 +1204,8 @@ def machine_i440FX(cmd=False):
                                              '_PCI_CHASSIS_NR', first_port=[1]),
                    qdevices.QCPUBus(params.get("cpu_model"), [[""], [0]],
                                     "vcpu"))
+            pflash_devices, cmd = pflash_handler("ovmf", cmd)
+            devices.extend(pflash_devices)
```

Comment:
Sorry @pevogam ,I missing this line before, to extend pflash devices.
After this fix, I tried to boot a guest with `pc` machine using OVMF, the result has a bit different from you.
```
2021-03-02 00:12:55,287 test             L0324 DEBUG|     machine_type = pc
2021-03-02 00:12:55,293 test             L0324 DEBUG|     ovmf_code_filename = OVMF_CODE.secboot.fd
2021-03-02 00:12:55,293 test             L0324 DEBUG|     ovmf_path = /usr/share/OVMF
2021-03-02 00:12:55,294 test             L0324 DEBUG|     ovmf_vars_filename = OVMF_VARS.fd
```

```
MALLOC_PERTURB_=1  /usr/libexec/qemu-kvm \
    -S  \
    -name 'avocado-vt-vm1'  \
    -sandbox on  \
    -blockdev node-name=file_ovmf_code,driver=file,filename=/usr/share/OVMF/OVMF_CODE.secboot.fd,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_ovmf_code,driver=raw,read-only=on,file=file_ovmf_code \
    -blockdev node-name=file_ovmf_vars,driver=file,filename=/var/lib/avocado/data/avocado-vt/images/rhel830-64-virtio-scsi.qcow2_VARS.fd,auto-read-only=on,discard=unmap \
    -blockdev node-name=drive_ovmf_vars,driver=raw,read-only=off,file=file_ovmf_vars \
    -machine pc,pflash0=drive_ovmf_code,pflash1=drive_ovmf_vars  \
    -nodefaults \
    -device VGA,bus=pci.0,addr=0x2 \
    -m 15360  \
    -smp 20,maxcpus=20,cores=10,threads=1,dies=1,sockets=2  \
    -cpu 'IvyBridge',+kvm_pv_unhalt \
```

The machine command provided by you is `-machine pc`, it lost `pflash0=drive_ovmf_code,pflash1=drive_ovmf_vars`, this is strange and I am not sure if your test case dropped OVMF variables before creating the guest.

Reply:
After your recent push, it now correctly adds the variables here too:
```
15:29:35 INFO | Running qemu command (reformatted):
/usr/bin/qemu-kvm \
    -S  \
    -name 'mis4' \
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/ovmf/OVMF_CODE.fd \
    -drive if=pflash,format=raw,file=/var/lib/avocado/data/avocado-vt/images/mis4_BusinessServer_64.qcow2_VARS.fd \
    -machine pc  \
    -nodefaults \
    -device virtio-vga,bus=pci.0,addr=0x2 \
    -m 1024  \
    -smp 2,maxcpus=2,cores=1,threads=1,dies=1,sockets=2  \
    -cpu 'EPYC' \
```
I still wonder why the variables are added right before the machine type and not together with the other drives but I guess the Qemu command line is not strictly ordered and this is not that important anyway.

Alright, so tests pass now. All we need is one more review and this can be merged :)

Reply:
OK, `-drive` mode will not append `pflash0=drive_ovmf_code,pflash1=drive_ovmf_vars` in the `-machine pc`, I forgot it.

This is in order to align with libvirt, add these two pflash devices to generate machine extra params at first, pass them into machine command.
```
/usr/libexec/qemu-kvm \
-name guest=avocado-vt-vm1,debug-threads=on \
-S \
-object secret,id=masterKey0,format=raw,file=/var/lib/libvirt/qemu/domain-1-avocado-vt-vm1/master-key.aes \
-blockdev {"driver":"file","filename":"/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw","node-name":"libvirt-pflash0-storage","auto-read-only":true,"discard":"unmap"} \
-blockdev {"node-name":"libvirt-pflash0-format","read-only":true,"driver":"raw","file":"libvirt-pflash0-storage"} \
-blockdev {"driver":"file","filename":"/var/lib/libvirt/qemu/nvram/avocado-vt-vm1_VARS.fd","node-name":"libvirt-pflash1-storage","auto-read-only":true,"discard":"unmap"} \
-blockdev {"node-name":"libvirt-pflash1-format","read-only":false,"driver":"raw","file":"libvirt-pflash1-storage"} \
-machine virt-rhel8.4.0,accel=kvm,usb=off,dump-guest-core=off,gic-version=3,pflash0=libvirt-pflash0-format,pflash1=libvirt-pflash1-format,memory-backend=mach-virt.ram
```